File: JavaScript/141-linked-list-cycle/index1.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *   this.val = val;
 *   this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
const hasCycle = function(head) {
  const linkSet = new Set();

  while (head !== null) {
    if (linkSet.has(head)) {
      return true;
    }

    linkSet.add(head);
    head = head.next;
  }

  return false;
};

File: Python/234-palindrome-linked-list/index1.py
# Definition for singly-linked list.
# class ListNode:
#   def __init__(self, x):
#     self.val = x
#     self.next = None

class Solution:
  def isPalindrome(self, head: ListNode) -> bool:
    prev = None
    slow = fast = head

    while fast and fast.next:
      fast = fast.next.next
      slow.next, prev, slow = prev, slow, slow.next

    if fast:
      slow = slow.next

    while prev and prev.val == slow.val:
      slow = slow.next
      prev = prev.next

    return not prev
